MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressHci.c

```diff
@@ -855,6 +872,139 @@ NvmeCreateIoSubmissionQueue (
+  UINT32      Index;
+  EFI_STATUS  Status;
+
+  // Offset completion queue with submission queue size
+  AsqsPages = EFI_SIZE_TO_PAGES (Private->SqData[1].NumberOfEntries * (UINTN)LShiftU64 (2, Private->SqData[1].EntrySize));
+
+  // Calculate the number of pages required for the admin queues
+  QueuePairPageCount = AsqsPages + EFI_SIZE_TO_PAGES (Private->CqData[1].NumberOfEntries * (UINTN)LShiftU64 (2, Private->CqData[1].EntrySize));
+
+  //
+  // Address of Data I/O submission & completion queue(s).
+  // We are using the same table of buffer pointers that the admin queus are in, so we start the table from Index + 1, but we have a separate
+  // buffer so we start at the beginning of that buffer.
+  //
+  ZeroMem (Private->DataQueueBuffer, EFI_PAGES_TO_SIZE (QueuePairPageCount) * Private->NumberOfDataQueuePairs);
+  for (Index = 0; Index < Private->NumberOfDataQueuePairs; Index++) {
+    Private->SqBuffer[Index + 1]        = (NVME_SQ *)(UINTN)(Private->DataQueueBuffer + Index * QueuePairPageCount * EFI_PAGE_SIZE);
+    Private->SqBufferPciAddr[Index + 1] = (NVME_SQ *)(UINTN)(Private->DataQueueBufferPciAddr + Index * QueuePairPageCount * EFI_PAGE_SIZE);
+    Private->CqBuffer[Index + 1]        = (NVME_CQ *)(UINTN)(Private->DataQueueBuffer + (Index * QueuePairPageCount + AsqsPages) * EFI_PAGE_SIZE);
+    Private->CqBufferPciAddr[Index + 1] = (NVME_CQ *)(UINTN)(Private->DataQueueBufferPciAddr + (Index * QueuePairPageCount + AsqsPages) * EFI_PAGE_SIZE);
+
+    DEBUG ((DEBUG_INFO, "Data IO   Submission Queue (SqBuffer[%d]) = [%016X]\n", Index + 1, Private->SqBuffer[Index + 1]));
+    DEBUG ((DEBUG_INFO, "Data IO   Completion Queue (CqBuffer[%d]) = [%016X]\n", Index + 1, Private->CqBuffer[Index + 1]));
+  }
+
+  //
+  // Create I/O completion queue(s).
+  //
+  Status = NvmeCreateIoCompletionQueue (Private);
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  //
+  // Create I/O Submission queue(s).
+  //
+  Status = NvmeCreateIoSubmissionQueue (Private);
+
+  return Status;
+}
+
+// MU_CHANGE [END]
+
+// MU_CHANGE [BEGIN] - Allocate IO Queue Buffer
+
+/**
+  Initialize the Nvm Express controller Admin Queues
+
+  @param[in] Private                 The pointer to the NVME_CONTROLLER_PRIVATE_DATA data structure.
+
+  @retval EFI_SUCCESS                The NVM Express Controller is initialized successfully.
+  @retval Others                     A device error occurred while initializing the controller.
+
+**/
+EFI_STATUS
+NvmeControllerInitAdminQueues (
+  IN NVME_CONTROLLER_PRIVATE_DATA  *Private
+  )
+{
+  NVME_ASQ    Asq;
+  NVME_ACQ    Acq;
+  UINTN       AsqsPages;
+  UINTN       QueuePairPageCount;
+  EFI_STATUS  Status;
+
+  // Offset completion queue with submission queue size
+  AsqsPages = EFI_SIZE_TO_PAGES (Private->SqData[0].NumberOfEntries * (UINTN)LShiftU64 (2, Private->SqData[0].EntrySize));
+
+  //
+  // Address of admin submission queue.
+  //
+  Asq = (UINT64)(UINTN)(Private->BufferPciAddr) & ~EFI_PAGE_MASK;

```

Comment:
Should page alignment be checked in the allocation function instead?

Reply:
Talked offline, PciIo->Map() uses AllocatePages so using this mask isn't necessary. Removing